src/render/ScreenCreditController.js

```diff
@@ -121,184 +108,33 @@ define([
-            gl.disableVertexAttribArray(program.vertexTexCoordLocation);
-
-            // Clear GL bindings.
-            gl.bindBuffer(gl.ARRAY_BUFFER, null);
-            gl.bindTexture(gl.TEXTURE_2D, null);
-
-            // Re-enable depth testing.
-            gl.enable(gl.DEPTH_TEST);
-        };
-
-        // Internal use only. Intentionally not documented.
-        ScreenCreditController.prototype.drawImageCredit = function (dc, creditUrl, x, y) {
-            var imageWidth, imageHeight, scale, activeTexture, gl, program;
-
-            activeTexture = dc.gpuResourceCache.resourceForKey(creditUrl);
-            if (!activeTexture) {
-                dc.gpuResourceCache.retrieveTexture(dc.currentGlContext, creditUrl);
-                return 0;
+        ScreenCreditController.prototype.doRender = function (dc) {
+            var creditOrdinal = 1;
+
+            for (var i = 0, len = this.imageCredits.length; i < len; i++) {
+                this.imageCredits[i].screenOffset.x = dc.viewport.width - (this.margin);
+                this.imageCredits[i].screenOffset.y = creditOrdinal * this.creditSpacing;
+                this.imageCredits[i].render(dc);
+                creditOrdinal++;
             }
 
-            // Scale the image to fit within a constrained size.
-            imageWidth = activeTexture.imageWidth;
-            imageHeight = activeTexture.imageHeight;
-            if (imageWidth <= this.imageCreditSize && this.imageHeight <= this.imageCreditSize) {
-                scale = 1;
-            } else if (imageWidth >= imageHeight) {
-                scale = this.imageCreditSize / imageWidth;
-            } else {
-                scale = this.imageCreditSize / imageHeight;
-            }
-
-            ScreenCreditController.imageTransform.setTranslation(x, y, 0);
-            ScreenCreditController.imageTransform.setScale(scale * imageWidth, scale * imageHeight, 1);
-
-            gl = dc.currentGlContext;
-            program = dc.currentProgram;
-
-            // Compute and specify the MVP matrix.
-            ScreenCreditController.scratchMatrix.copy(dc.screenProjection);
-            ScreenCreditController.scratchMatrix.multiplyMatrix(ScreenCreditController.imageTransform);
-            program.loadModelviewProjection(gl, ScreenCreditController.scratchMatrix);
-
-            program.loadTextureEnabled(gl, true);
-            program.loadColor(gl, Color.WHITE);
-            program.loadOpacity(gl, this.opacity);
-
-            ScreenCreditController.texCoordMatrix.setToIdentity();
-            ScreenCreditController.texCoordMatrix.multiplyByTextureTransform(activeTexture);
-            program.loadTextureMatrix(gl, ScreenCreditController.texCoordMatrix);
-
-            if (activeTexture.bind(dc)) { // returns false if active texture cannot be bound
-                // Draw the image quad.
-                gl.drawArrays(gl.TRIANGLE_STRIP, 0, 4);
+            for (i = 0; i < this.textCredits.length; i++) {
```

Comment:
Two problems here:
• Array.length is called every iteration.
• Variable i should be declared at the top of the method. Avoid dependencies on variable hosting.

Reply:
Whoops on that Array.length call.
Both points are corrected now.